### ISSUE TITLE
Referencing files in resources/jar

### ISSUE THREAD

**Issue by antonioalegria** [eclipsesource/play-json-schema-validator#48]
Hi,

I'm having trouble referencing ($ref) other schema files that sit together inside a jar/resources.

I've tried using:

- "$ref": "resource-name.json" -> File not found exception
- "$ref": "file:resource-name.json" -> File not found exception
- "$ref": "file:./resource-name.json" -> File not found exception
- "$ref": "resource:resource-name.json" -> Malformed URL
- "$ref": "resource:/resource-name.json" -> Malformed URL
- "$ref": "classpath:resource-name.json" -> Malformed URL
- "$ref": "classpath:/resource-name.json" -> Malformed URL

I'm running out of ideas. Is there something obvious I'm forgetting or is this something not supported by your library?

Thank you!

**Comment by edgarmueller** [comment]
Hi,
correct me if I am wrong, but I believe this is not possible in general, i.e. relatively referencing files within a JAR. `resource` and `classpath` are invalid protocols (classpath is Spring-specific, I think) and the `file` protocol can not be used to reference files within a JAR file. Also, you can't use the `file` protocol with a relative URL and you can't know the absolute path to the file within the JAR when writing the schema either (well, you can, but that is probably a bad idea, since you would need to know the resource runtime path in advance).
Could you elaborate your use case a bit more? Perhaps extracting the JAR and just using relative refs would be option?
Thanks!

**Comment by antonioalegria** [comment]
From what I've read around in the web, https://github.com/fge/json-schema-validator supports this via the classpath or reference protocols. They probably register a protocol handler for this.

This seems to be a very common use case – it would be great if I could have schemas referenced within a packaged JAR.

For now I have to do what you suggested, which is certainly not optimal.

Thank you!

**Comment by edgarmueller** [comment]
Ok, I see what I can come up with for the next release, thanks for letting me know.

**Comment by avkonst** [comment]
I have seen other library (could not find now) which supported a concept of UrlResolver. An instance of this class is supplied to schema validator instance.. Schema validator invokes UrlResolver whenever it detects a reference. UrlResolver can implement custom file fetch mechanisms and recognize non-standard protocols. UrlResolver had a single function "resolve(urlFilePath: string) -> fileContent: string".

**Comment by edgarmueller** [comment]
I've merged a change that enables a custom classpath protocol, [here's](https://github.com/eclipsesource/play-json-schema-validator/blob/master/src/test/scala/com/eclipsesource/schema/SchemaValidatorSpec.scala#L168) the respective test case. @antonioalegria Does this cover your use case? `some.json` is located within a JAR on the classpath and references `loc.json' within the same JAR via the following reference:
```
{
    "location": { "$ref": "classpath:/loc.json" }
}
```
@avkonst Thanks for the idea. The API changed slightly as the validator isn't a object anymore but a class, since it holds some state about the available URL resolvers. A [`URLResolver`](https://github.com/eclipsesource/play-json-schema-validator/blob/master/src/main/scala/com/eclipsesource/schema/internal/url/UrlResolver.scala) is nothing more than a `URLStreamHandler` with a `protocol` member.

**Comment by edgarmueller** [comment]
Fixed in 0.8.0

**Comment by andrewmeyerBR** [comment]
What about?:
```
{
    "location": { "$ref": "classpath:loc.json" }
}
```

Note the lack of `/`. It seems relative references are not supported, even with the `ClasspathUrlResolver`. Is this a correct statement?
